multicluster/controllers/multicluster/gateway_controller.go

```diff
@@ -0,0 +1,277 @@
+	commonArea, err := r.getCommonArea()
+	if commonArea == nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	r.leaderNamespace = commonArea.GetNamespace()
+	err = r.getServiceCIDR(ctx)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	resExportName := newClusterInfoResourceExportName(r.localClusterID)
+	resExportNamespacedName := types.NamespacedName{
+		Name:      resExportName,
+		Namespace: r.leaderNamespace,
+	}
+	resExport := &mcsv1alpha1.ResourceExport{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      resExportName,
+			Namespace: r.leaderNamespace,
+		},
+	}
+
+	createOrUpdate := func(gwIP string) error {
+		existingResExport := &mcsv1alpha1.ResourceExport{}
+		if err := commonArea.Get(ctx, resExportNamespacedName, existingResExport); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			if err = r.createResourceExport(ctx, req, commonArea, existingResExport, gwIP); err != nil {
+				return err
+			}
+			return nil
+		}
+		if err = r.updateResourceExport(ctx, req, commonArea, existingResExport, gwIP); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	gw := &mcsv1alpha1.Gateway{}
+	if err := r.Client.Get(ctx, req.NamespacedName, gw); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+		gwInfo, err := r.getLastCreatedGateway()
+		if err != nil {
+			klog.ErrorS(err, "Failed to get Gateways")
+			return ctrl.Result{}, err
+		}
+		if gwInfo == nil {
+			// When the last Gateway is deleted, we will remove the ClusterInfo kind of ResourceExport
+			if err := commonArea.Delete(ctx, resExport, &client.DeleteOptions{}); err != nil {
+				return ctrl.Result{}, client.IgnoreNotFound(err)
+			}
+			return ctrl.Result{}, nil
+		}
+		// When there are still Gateways exist, we should create or update existing ResourceExport
+		// with the latest Gateway in remaining Gateways.
+		if err := createOrUpdate(gwInfo.GatewayIP); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+	if err := createOrUpdate(gw.GatewayIP); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+// getCommonArea checks ClusterSet initialization status. When ClusterSet is initialized successfully by member
+// ClusterSet controller, remoteCommonAreaManager will be built and include a connected CommonArea.
+func (r *GatewayReconciler) getCommonArea() (commonarea.RemoteCommonArea, error) {
+	remoteCommonAreaManager := r.mcReconciler.GetRemoteCommonAntreaManager()
+	if remoteCommonAreaManager == nil {
+		klog.InfoS("ClusterSet has not been initialized properly, no available common area")
+		return nil, nil
+	}
+	r.localClusterID = string((remoteCommonAreaManager).GetLocalClusterID())
+	if len(r.localClusterID) == 0 {
+		klog.InfoS("localClusterID is not initialized, retry later")
+		return nil, nil
+	}
+
+	commonArea, err := getRemoteCommonArea(remoteCommonAreaManager)
+	if err != nil {
+		return nil, err
+	}
+	return commonArea, nil
```

Comment:
The RemoteCommonAreaManager is stopped only when the ClusterSet is deleted https://github.com/antrea-io/antrea/blob/6c704e3c2f03ff03b6c322768e0bd9937ae29dab/multicluster/controllers/multicluster/member_clusterset_controller.go#L74, it will be back to the state just like there is no ClusterSet, but we can tell if RemoteCommonAntreaManager is stopped or it's not initialized in the log.
When it's stopped, we will see error like `ClusterSet has not been initialized properly, no available common area`, and all up-coming event will be requeued due to this error until ClusterSet is initialized correctly again.

Reply:
Probably add a comment after getting RemoteCommonAntreaManager saying it can be stopped, and then all CommonArea calls will fail with an error (is it true?).

Other two questions: 1) have we implemented reelection of leader cluster when a leader is down? 2) when a ClusterSet is created/re-created, I did not see we reconcile all existing Gateways or SericeExports?

Reply:
I think all the events after RemoteCommonAntreaManager  is stopped will fail, but if the RemoteCommonAntreaManager is stopped during a reconciling process, for example, controller is in the middle of process a Service update event, then it may still work because of ClusterSet clear up(off boarding) process is not fully implemented, eg: RBAC in leader is still existing, secret token for member cluster to access leader is also not deleted.

Reply:
For the two questions:
1. yes, @aravindakidambi implemented the reelection process, when leader controller is restarted, it will reelect(reconnect) the leader cluster as common area.
2. All existing Gateways or ServiceExports will be requeued and wait until there is a valid RemoteCommonAntreaManager, so they will be reconciled once a comon area is connected.

Reply:
I feel this way is not correct if I understood you correctly. We should create ResourceExports for all ServiceExports and Gateways when a new leader is elected, not just the changed ones.

Reply:
OK, I got your point, but we haven't implemented anything about Leader HA for resource exchange, we assume that there is only one leader for now, the reelection is actually from disconnected to connected status. If we plan to support leader HA, it might be totally different for resource exchange. I will add a TODO for it if we do have any plan for it.

Reply:
Then what you mean by "RemoteCommonAreaManager is stopped only when the ClusterSet is deleted"? Do we support re-creating/re-joining a ClusterSet?

And what you mean by "implemented the reelection process, when leader controller is restarted, it will reelect(reconnect) the leader cluster as common area"? We implemented reelection but it is actually dead code with no use?

Reply:
@luolanzone : could you answer my questions above?

Reply:
@luolanzone : could you answer my questions above?

Reply:
sorry, missed this question, Current codes doesn't support to rejoin a different ClusterSet, but it should be OK to rejoin to existing ClustserSet if ClusterSet is deleted in member cluster only. We didn't support multiple Leader clusters HA for resource exchange since we can't guarantee all member clusters will join to the same leader clusters in the same time, which means we need to make sure data replication in the Leader cluster. As we discussed, I will revisit this part to see if we need to remove dead code or refine them to support HA.